error c3203

by adding this 2 lines, should I have import the module?
#include "xtensor\xarray.hpp"
#include "xtensor\xio.hpp"

after so , I got 
c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(648): error C3203: “layout_iterator”
c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(674): error C3203: “const_layout_iterator'
c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(767): error C3203: “const_reverse_layout_iterator”

is there anything wrong I have done?

THX

I didn't declare any variable in the code.
I just adding
```
#include "xtensor\xarray.hpp"
#include "xtensor\xio.hpp"
```
at the front and begin compiling.
then I got 
```
1>------ Build started: Project: TDFEasyDemo27, Configuration: Debug Win32 ------
1>TDFEasyDemo.cpp
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(648): error C3203: 'layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(661): error C3203: 'layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(674): error C3203: 'const_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(681): error C3203: 'const_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(688): error C3203: 'const_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(701): error C3203: 'const_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(714): error C3203: 'reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(727): error C3203: 'reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(740): error C3203: 'const_reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(747): error C3203: 'const_reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(754): error C3203: 'const_reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>c:\users\zczy4\desktop\c\include\xtensor\xcontainer.hpp(767): error C3203: 'const_reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>Done building project "TDFEasyDemo27.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

```

Hi @SylvainCorlay, I created the [xp-xtensor](https://github.com/roliveira/xp-xtensor) repository to test those issues out as the original project I was using was a little bit messy...

I'm building it using CMake with the Visual Studio 15 2017 Win64 generator (v141), nothing fancy just the first basic usage example, thought I get the same by simply including the headers (like @tanruoqiao). I'm getting loads of those C3203 errors:

```
1>------ Build started: Project: xp-xtensor, Configuration: Release x64 ------
2>------ Build started: Project: xtensor, Configuration: Release x64 ------
3>------ Build started: Project: xtl, Configuration: Release x64 ------
1>Building Custom Rule D:/GitHub/xp-xtensor/CMakeLists.txt
3>Performing update step for 'xtl'
2>Performing update step for 'xtensor'
1>CMake does not need to re-run because D:/GitHub/xp-xtensor/build/CMakeFiles/generate.stamp is up-to-date.
1>main.cpp
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(690): error C3203: 'layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(703): error C3203: 'layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(716): error C3203: 'const_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(723): error C3203: 'const_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(730): error C3203: 'const_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(743): error C3203: 'const_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(756): error C3203: 'reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(769): error C3203: 'reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(782): error C3203: 'const_reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(789): error C3203: 'const_reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(796): error C3203: 'const_reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>d:\github\xp-xtensor\build\xtensor\src\xtensor\include\xtensor\xcontainer.hpp(809): error C3203: 'const_reverse_layout_iterator': unspecialized alias template can't be used as a template argument for template parameter '_Ty2', expected a real type
1>Done building project "xp-xtensor.vcxproj" -- FAILED.
3>Already up-to-date.
2>Already up-to-date.
4>------ Skipped Build: Project: ALL_BUILD, Configuration: Release x64 ------
4>Project not selected to build for this solution configuration 
========== Build: 2 succeeded, 1 failed, 1 up-to-date, 1 skipped ==========
```

Any thoughts on it?

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -20,7 +20,7 @@ init:
   - if "%APPVEYOR_BUILD_WORKER_IMAGE%" == "Visual Studio 2015" set VCVARPATH="C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat"
   - if "%APPVEYOR_BUILD_WORKER_IMAGE%" == "Visual Studio 2015" set VCARGUMENT=%PLATFORM% 
   - if "%APPVEYOR_BUILD_WORKER_IMAGE%" == "Visual Studio 2017" if "%PLATFORM%" == "x64" set VCVARPATH="C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvars64.bat"
-  - if "%APPVEYOR_BUILD_WORKER_IMAGE%" == "Visual Studio 2017" if "%PLATFORM%" == "x32" set VCVARPATH="C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvars32.bat"
+  - if "%APPVEYOR_BUILD_WORKER_IMAGE%" == "Visual Studio 2017" if "%PLATFORM%" == "x86" set VCVARPATH="C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvars32.bat"
   - echo "%VCVARPATH% %VCARGUMENT%"
   - "%VCVARPATH% %VCARGUMENT%"
   - ps: if($env:Platform -eq "x64"){Start-FileDownload 'http://repo.continuum.io/miniconda/Miniconda3-latest-Windows-x86_64.exe' C:\Miniconda.exe; echo "Done"}

--- a/docs/source/compilers.rst
+++ b/docs/source/compilers.rst
@@ -14,6 +14,11 @@ Visual Studio 2015 and ``std::enable_if``
 
 With Visual Studio, ``std::enable_if`` evaluates its second argument, even if the condition is false. This is the reason for the presence of the indirection in the implementation of the ``xfunction_type_t`` meta-function.
 
+Visual Studio 2017 and alias templates with non-class template parameters and multiple aliasing levels
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Alias template with non-class parameters only, and multiple levels of aliasing are not properly considered as types by Visual Studio 2017. The base ``xcontainer`` template class underlying xtensor container types has such alias templates defined. We avoid the multiple levels of aliasing in the case of Visual Studio.
+
 GCC-4.9 and Clang < 3.8 and constexpr ``std::min`` and ``std::max``
 -------------------------------------------------------------------
 

--- a/include/xtensor/xcontainer.hpp
+++ b/include/xtensor/xcontainer.hpp
@@ -143,6 +143,17 @@ namespace xt
         template <class align, class simd = simd_value_type>
         simd load_simd(size_type i) const;
 
+#if defined(_MSC_VER) && _MSC_VER >= 1910
+        // Workaround for compiler bug in Visual Studio 2017 with respect to alias templates with non-type parameters.
+        template <layout_type L>
+        using layout_iterator = xiterator<typename iterable_base::stepper, typename iterable_base::inner_shape_type*, L>;
+        template <layout_type L>
+        using const_layout_iterator = xiterator<typename iterable_base::const_stepper, typename iterable_base::inner_shape_type*, L>;
+        template <layout_type L>
+        using reverse_layout_iterator = std::reverse_iterator<layout_iterator<L>>;
+        template <layout_type L>
+        using const_reverse_layout_iterator = std::reverse_iterator<const_layout_iterator<L>>;
+#else
         template <layout_type L>
         using layout_iterator = typename iterable_base::template layout_iterator<L>;
         template <layout_type L>
@@ -151,6 +162,7 @@ namespace xt
         using reverse_layout_iterator = typename iterable_base::template reverse_layout_iterator<L>;
         template <layout_type L>
         using const_reverse_layout_iterator = typename iterable_base::template const_reverse_layout_iterator<L>;
+#endif
 
         template <class S, layout_type L>
         using broadcast_iterator = typename iterable_base::template broadcast_iterator<S, L>;

